### PR TITLE
Performance optimization

### LINKED ISSUE #10: File transfer example
It would be nice to have an example for how to best transfer a file to/from the remote machine.
e.g. a function like `push_file(source_path: &str, dest_path: &str)` that pushes from the local machine to the remote machine.

Doing this efficiently would be complicated since I imagine we would want to read new bytes from the source while sending to the destination?
Additionally this is a very common use case for sftp, so I think making an example of this would bring a lot of value.

### PR DESCRIPTION
**Pull request by zidaye**
close #10 

```sh
# run
cargo bench --bench my_benchmark -- --verbose
```
#### 8 files, each with a size of 10 MB, uploaded in approximately 3.3 seconds.  The speed is 24 MB/s.  
image

#### In fact, by [modifying the code to increase concurrency](https://github.com/warp-tech/russh/blob/271e362bdebc994aafaf340fb845d20125d0db09/russh/src/client/mod.rs#L671C15-L671C15), it is possible to achieve an upload speed of 30-40 MB/s. 
like this:
```rust
let (handle_sender, session_receiver) = channel(64);
```

### CODE REVIEW

**Review by AspectUnk — APPROVED** [review]

File: src/client/rawsession.rs [hunk 1]
```diff
@@ -128,10 +123,11 @@ pub(crate) struct Options {
 /// If the server returns a `Status` packet and it has the code Ok
 /// then the packet is returned as Ok in other error cases
 /// the packet is stored as Err.
+#[derive(Debug, Clone)]
 pub struct RawSftpSession {
     tx: mpsc::UnboundedSender<Bytes>,
     requests: Arc<SharedData>,
-    next_req_id: u32,
+    next_req_id: Arc<Mutex<u32>>,
     handles: Wrapping<u64>,
```
Comment (AspectUnk) [inline src/client/rawsession.rs:131]: The value will not be synchronized when cloned

File: src/client/rawsession.rs [hunk 2]
```diff
@@ -128,10 +123,11 @@ pub(crate) struct Options {
 /// If the server returns a `Status` packet and it has the code Ok
 /// then the packet is returned as Ok in other error cases
 /// the packet is stored as Err.
+#[derive(Debug, Clone)]
```
Comment (AspectUnk) [inline src/client/rawsession.rs:126]: Cloning creates a lot of internal value synchronization problems

File: src/client/rawsession.rs [hunk 3]
```diff
@@ -191,24 +187,28 @@ impl RawSftpSession {
...
                 Err(error.into())
             }
         }
     }
 
-    fn use_next_id(&mut self) -> u32 {
-        let id = self.next_req_id;
-        self.next_req_id += 1;
-        id
+    async fn use_next_id(&mut self) -> u32 {
```
Comment (AspectUnk) [inline src/client/rawsession.rs:207]: Can be atomic

File: src/client/rawsession.rs (hunk 1 above)
Comment (AspectUnk) [inline src/client/rawsession.rs:131]: Handles also do not have synchronization

**Comment by AspectUnk** [comment]
Thanks for taking the time to do this. This implementation has several synchronization problems, including handles, and the `request_id` itself can be atomic, but I will fix all this myself in another branch.

**Comment by UVUUUVUU** [comment]
> Thanks for taking the time to do this. This implementation has several synchronization problems, including handles, and the `request_id` itself can be atomic, but I will fix all this myself in another branch.

This has been very helpful in my work and I am very much looking forward to this part of the feature

**Pull request merged**